penaltymodel_lp/penaltymodel/lp/generation.py

```diff
@@ -0,0 +1,137 @@
+
+    # Populate linear terms (i.e. spin states)
+    matrix[:, :m_linear] = spin_states
+
+    # Populate quadratic terms
+    node_indices = dict(zip(nodes, range(m_linear)))
+    for j, (u, v) in enumerate(edges):
+        u_ind = node_indices[u]
+        v_ind = node_indices[v]
+        matrix[:, j + m_linear] = np.multiply(matrix[:, u_ind], matrix[:, v_ind])
+
+    # Populate offset and gap columns, respectively
+    matrix[:, -2] = offset_weight
+    matrix[:, -1] = gap_weight
+    return matrix
+
+
+#TODO: check table is not empty (perhaps this check should be in bqm.stitch or as a common
+# penaltymodel check)
+#TODO: Check if len(table.keys()[0]) == len(decision_variables)
+
+#TODO: Adding min_gap into code
+#TODO: Checking min_gap < max_gap (occurs at bqm.stitch or penaltymodel?)
+def generate_bqm(graph, table, decision_variables,
+                 linear_energy_ranges=None, quadratic_energy_ranges=None):
+
+    # Check for auxiliary variables in the graph
+    if len(graph) != len(decision_variables):
+        raise ValueError('Penaltymodel-lp does not handle problems with auxiliary variables')
+
+    if not linear_energy_ranges:
+        linear_energy_ranges = {}
+
+    if not quadratic_energy_ranges:
+        quadratic_energy_ranges = {}
+
+    # Simplify graph naming
+    # Note: nodes' and edges' order determine the column order of the LP
+    nodes = decision_variables
+    edges = graph.edges
+
+    # Set variable names for lengths
+    m_linear = len(nodes)                   # Number of linear biases
+    m_quadratic = len(edges)                # Number of quadratic biases
+    n_noted = len(table)                    # Number of spin combinations specified in the table
+    n_unnoted = 2**m_linear - n_noted       # Number of spin combinations that were not specified
+
+    # Linear programming matrix for spin states specified by 'table'
+    noted_states = table.keys() if isinstance(table, dict) else table
+    noted_states = list(noted_states)
+    noted_matrix = _get_lp_matrix(np.asarray(noted_states), nodes, edges, 1, 0)
+
+    # Linear programming matrix for spins states that were not specified by 'table'
+    spin_states = product([-1, 1], repeat=m_linear)
+    unnoted_states = [state for state in spin_states if state not in noted_states]
+    unnoted_matrix = _get_lp_matrix(np.asarray(unnoted_states), nodes, edges, 1, -1)
+    if unnoted_matrix is not None:
+        unnoted_matrix *= -1   # Taking negative in order to flip the inequality
+
+    # Constraints
+    if isinstance(table, dict):
+        noted_bound = np.asarray([table[state] for state in noted_states])
+        unnoted_bound = np.full((n_unnoted, 1), -1 * min(table.values()))  # -1 for flipped inequality
+    else:
+        noted_bound = np.zeros((n_noted, 1))
+        unnoted_bound = np.zeros((n_unnoted, 1))
+
+    # Bounds
+    max_gap = m_linear * MAX_LINEAR_BIAS + m_quadratic * MAX_QUADRATIC_BIAS
```

Comment:
Probably want to determine this from the energy ranges, defaulting to MAX_LINEAR_BIAS